### PR TITLE
Add support for stm32f446re chip and nucleo-f446re board

### PR DESCRIPTION
**Pull request by rajivr**
### Pull Request Overview

This pull request adds Tock support for STM32F446RE based [NUCLEO-F446RE](https://www.st.com/en/evaluation-tools/nucleo-f446re.html) development board.

I am hoping we can review this port in 1.4 development cycle and if its acceptable to the core team, consider it for inclusion in upstream tree.

Following STM32F446RE peripherals and Tock capsules are enabled.

#### STM32F446RE peripherals

- RCC
- GPIO
- DMA1
- USART2
- EXTI
- SYSCFG
- TIM2
- DBG

#### Tock capsules

- Console
- LED
- Button
- Alarm

This port introduces some new ideas and patterns and I'll summarize them below.

~~#### Separation of SVD generated files from driver source~~

~~Files generated using `svd2regs.py` tool is labeled as `svd_XXX.rs`, and I am using `include!("svd_XXX.rs")` macro to pull the register definition into the driver code.~~

~~#### Usage of Events and Event Dispatchers~~

~~When I started learning about Tock interrupt handling mechanism, I [observed](https://groups.google.com/forum/#!topic/tock-dev/fZQYq0dpeSQ) that there was a possibility for lost interrupts. Then in #1181 @lthiery introduced the idea of event dispatch mechanism for interrupts. @phil-levis [added](https://groups.google.com/d/msg/tock-dev/vWkRLr_A47g/x3iO8tHrBAAJ) additional things that the new design should address.~~

~~Following is an overview of how event dispatch mechanism works in this port.~~

~~1. The biggest change is that I am no longer disabling interrupts at NVIC in the interrupt handler. You'll notice, I don't use `generic_isr!` for the driver top-half interrupt handler. In a nutshell, I was unable to reason about correctness, when interrupts were disabled at NVIC.~~

~~2. Building upon @lthiery's work, I introduce two macros `interrupt_context_isr!` and `kernel_context_isr!`. This provides mechanisms for handling interrupts in interrupt context and kernel context. With `interrupt_context_isr!`, the need for disabling interrupts at NVIC doesn't arise, because we can take care of dealing with peripheral interrupt source in interrupt context (top-half), and do the remaining interrupt processing in kernel context (bottom-half).~~

~~3. All drivers that require interrupt handling (DMA1, USART2, EXTI, TIM2) use `interrupt_context_isr!`. Also, most vendor HAL's are written with some form of processing that is done in interrupt context, so having `interrupt_context_isr!` helps us learn from vendor HAL code when developing new Tock drivers.~~

~~4. The external interrupt controller (EXTI) on STM32F446RE, muxes some GPIO interrupts to the same line on NVIC. This required introduction of `EXTI_EVENTS`, in addition to `INTERRUPT_EVENTS`.~~

~~The Chip trait also needed some modifications, which makes it a breaking change for now. My hope is to start a conversation about interrupt handling design in Tock along with @lthiery's #1181 pull-request and @phil-levis proposal.~~

### Testing Strategy

This pull request was tested using following apps in `libtock-c`

- `examples/blink`
- `examples/button`
- `examples/c_hello`
- `examples/tests/console_timeout`
- `examples/tests/console`
- `examples/tests/hello_loop`

### TODO or Help Wanted

Getting interrupt handling right can be tricky due to subtle edge cases. So, please poke holes. 

This is also my first pull request to Tock, so if I've missed any aspects of the contribution process, please let me know.

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make formatall`.

### CODE REVIEW

**Comment by bradjc** [comment]
OK this is awesome. Sorry you only have a reaction so far! We are right in the middle of a release (https://github.com/tock/tock/issues/1201) but after that is settled we will definitely start merging PRs again!

File: boards/kernel_layout.ld [hunk 1]
```diff
@@ -137,6 +137,19 @@ SECTIONS
...
+        /* _sapps symbol used by tock to look for first application */
+        . = ALIGN(4);
+        _sapps = .;
+
+	LONG(0)
+
+        /* Optional .app sections a convenience mechanism to bundle tock
+           kernel and apps into a single image */
+        KEEP (*(.app.*))
+    } > prog
```
Comment (bradjc) [inline boards/kernel_layout.ld]: Did moving this and adding `LONG(0)` fix an issue you were having?

File: boards/nucleo_f446re/Makefile [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Makefile for building the tock kernel for the NUCLEO-446RE platform
+#
+TARGET=thumbv7em-none-eabi
+PLATFORM=nucleo_f446re
+
+include ../Makefile.common
```
Comment (bradjc) [inline boards/nucleo_f446re/Makefile]: How are you loading code onto the board?

File: kernel/src/syscall.rs [hunk 3]
```diff
@@ -132,3 +132,57 @@ pub trait UserspaceKernelBoundary {
         writer: &mut Write,
     );
 }
+
+/// A dummy `UserspaceKernelBoundary` implementation, used to satisfy the type
+/// system when bootstrapping a new arch/chip crate.
+impl UserspaceKernelBoundary for () {
```
Comment (bradjc) [inline kernel/src/syscall.rs]: The cortex-m `UserspaceKernelBoundary` should just work, shouldn't it?

I think adding a new arch is so rare that this wouldn't get used in practice, and anything that needs it doesn't really support Tock enough to be merged into mainline.

**Comment by rajivr** [comment]
Congrats to you and rest of the Tock core team on 1.3 release.

Regarding SVD split, following is what I had in mind on how we could abstract across MCU families. 

1. If the difference between MCUs was only with the MMIO base address, then we could use `#[cfg(...)]` to include the appropriate SVD file.

2. If there was additional differences, then we could introduce a module that would abstract over those differences, and within that module we can do `#[(cfg(...)]` and `#include!(...)`.

I am okay with including the `svd_XXX.rs` register definitions into the driver code, but I wanted to share the motivation behind the split.

File: boards/kernel_layout.ld (hunk 1 above)
Comment (rajivr) [reply boards/kernel_layout.ld]: I initially removed `.apps :` in order to compile and test `kernel_loop` without `load_process`. However when I put `.apps :` back, I put it at a different place in `kernel_layout.ld`. Hence it showed up as a move. I'll correct it in the next push.

Regarding `LONG(0)`, this is needed for `arm-none-eabi-objcopy --update-section` to work properly. Without `LONG(0)`, `arm-none-eabi-objcopy` gives the following error.

```
arm-none-eabi-objcopy:target/thumbv7em-none-eabi/debug/nucleo_f446re-app[.apps]: Section has no contents
```

File: boards/nucleo_f446re/Makefile (hunk 2 above)
Comment (rajivr) [reply boards/nucleo_f446re/Makefile]: I am using `arm-none-eabi-objcopy --update-section` to create an ELF image with the apps.

```diff
diff --git a/boards/Makefile.common b/boards/Makefile.common
index c76f2422..24941be9 100644
--- a/boards/Makefile.common
+++ b/boards/Makefile.common
@@ -172,3 +172,4 @@ target/$(TARGET)/release/$(PLATFORM):
 target/$(TARGET)/debug/$(PLATFORM):
        $(Q)RUSTFLAGS="$(RUSTFLAGS_FOR_CARGO_LINKING)" $(CARGO) build $(VERBOSE) --target=$(TARGET)
        $(Q)$(SIZE) $@
+       arm-none-eabi-objcopy --update-section .apps=/home/tos-user/work/src/libtock-c/examples/buttons/build/cortex-m4/cortex-m4.tbf $@ $@-app
```
Then using gdb/openocd to load `nucleo_f446re-app` onto the board.

**Comment by bradjc** [comment]
> If the difference between MCUs was only with the MMIO base address, then we could use #[cfg(...)] to include the appropriate SVD file.

Wouldn't all of the actual driver implementation code be shared too? I ran into this issue with the RISC-V boards from SiFive, and came up with this approach: https://github.com/tock/tock/blob/riscv/chips/arty_exx/src/pwm.rs where the shared code lives in a crate, and the chip-specific code essentially just calls the `new()` function.

> If there was additional differences, then we could introduce a module that would abstract over those differences, and within that module we can do #[(cfg(...)] and #include!(...).

Other collaborators may not feel as strongly, but I am very opposed to cfg statements, which are just `#ifdef` by another name, and make reading code bases extremely difficult, particularly for new developers/users. But my thoughts on that are really neither here nor there, and we can deal with any cases like this on a case-by-case basis.


My preference is to keep related code together, and keep consistency with other boards (some of which had separate files and have been slowly consolidated).

File: kernel/src/syscall.rs (hunk 3 above)
Comment (bradjc) [reply kernel/src/syscall.rs]: That seems fine to me? If it is a new chip with a supported architecture, then it makes sense to use it (and if there are no processes on the board the code won't get called anyway). If it is a new chip with an unsupported arch then creating an empty `UserspaceKernelBoundary` implementation is useful because it will need to get filled in eventually anyway to be merged into Tock.

File: boards/nucleo_f446re/Makefile (hunk 2 above)
Comment (bradjc) [reply boards/nucleo_f446re/Makefile]: Can you add those commands to the Makefile?

**Comment by rajivr** [comment]
> My preference is to keep related code together, and keep consistency with
> other boards (some of which had separate files and have been slowly
> consolidated).

Okay. I'll merge the files in the next push.

File: boards/nucleo_f446re/Makefile (hunk 2 above)
Comment (rajivr) [reply boards/nucleo_f446re/Makefile]: I can add `flash` and `flash-debug` targets to `Makefile`. 

However, adding `program` target is going to be tricky for two reasons.

1. There is no tockloader support for this board.

2. Developing tockloader support is going to be challenging because the sector
   erase size is either 16KB, 64KB or 128KB. In hindsight, I feel I should have
   chosen a more modern STM32 board that supports 4KB or 2KB page erase.
         
In the `README.md`, I can document how to load programs onto the board, along
with a commented `program` target in `Makefile`.

Comment (bradjc) [reply boards/nucleo_f446re/Makefile]: Just `flash` is fine, as long as there is some known way to load code.

And re: erase size, what does it depend on? 16K is pretty big, and we'd actually have to make some improvements to tockloader-proto, but it _should_ work (even if it won't be memory efficient...)

Comment (rajivr) [reply boards/nucleo_f446re/Makefile]: > And re: erase size, what does it depend on? 16K is pretty big, and we'd
> actually have to make some improvements to tockloader-proto, but it _should_
> work (even if it won't be memory efficient...)

Erase size depends on MCU's flash memory controller. Some MCU's specify a page
erase, while others specify a sector erase.

In case of STM32F446RE, the sectors are organized as follows.

```
+----------+--------------------------+--------+
| Sector 0 | 0x0800_0000 - 0x0800_3FFF |  16KB |
+----------+---------------------------+-------+
| Sector 1 | 0x0800_4000 - 0x0800_7FFF |  16KB |
+----------+---------------------------+-------+
| Sector 2 | 0x0800_8000 - 0x0800_BFFF |  16KB |
+----------+---------------------------+-------+
| Sector 3 | 0x0800_C000 - 0x0800_FFFF |  16KB |
+----------+---------------------------+-------+
| Sector 4 | 0x0801_0000 - 0x0801_FFFF |  64KB |
+----------+---------------------------+-------+
| Sector 5 | 0x0802_0000 - 0x0803_FFFF | 128KB |
+----------+---------------------------+-------+
| Sector 6 | 0x0804_0000 - 0x0805_FFFF | 128KB |
+----------+---------------------------+-------+
| Sector 7 | 0x0806_0000 - 0x0807_FFFF | 128KB |
+----------+---------------------------+-------+
```

Hence we can only erase in blocks of either 16KB, 64KB or 128KB.

**Comment by rajivr** [comment]
I just pushed an updated patch series, with the following changes.

- Use `tock-rt0`
- Remove `dsb` instruction support
- Reorganize `kernel_layout.ld`
- Rebase to master
- Remove `svd_XXX.rs` files
- Add `flash` and `flash-debug` targets to `Makefile`
- Document `program` target in `README.md`

Please review.

**Comment by rajivr** [comment]
I just pushed an updated patch series, with the following changes.

- Use `tock-rt0`
- Remove `dsb` instruction support
- Reorganize `kernel_layout.ld`
- Rebase to master
- Remove `svd_XXX.rs` files
- Add `flash` and `flash-debug` targets to `Makefile`
- Document `program` target in `README.md`

Please review.

File: boards/nucleo_f446re/Makefile (hunk 2 above)
Comment (bradjc) [reply boards/nucleo_f446re/Makefile]: Hm, that does make things a little complicated. I think it could be done (since writes don't have to be that large) but it would likely take a few tweaks.

File: boards/kernel_layout.ld [hunk 4]
```diff
@@ -191,6 +191,8 @@ SECTIONS
         . = ALIGN(4);
         _sapps = .;
 
+	LONG(0)
```
Comment (bradjc) [inline boards/kernel_layout.ld]: I think we need a comment on this so that future linker script editors know why it is there.

Comment (ppannuto) [reply boards/kernel_layout.ld]: Also, does this break the bundling mechanism right now? I think this needs to come after the `KEEP(*(.app*))` below

File: boards/nucleo_f446re/Makefile (hunk 2 above)
Comment (ppannuto) [reply boards/nucleo_f446re/Makefile]: re `program`, please do add an empty target with explanation, similar to https://github.com/tock/tock/blob/master/boards/nordic/nrf51dk/Makefile#L27

We try to ensure that both `make program` and `make flash` "work", even if it's only meaningful to print directions on how to do the other thing in some cases

File: chips/stm32f446re/src/chip.rs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        //    this involves clearing SCB' deepsleep bit to prevent us from
+        //    accidentally going to deep sleep. In normal sleep, its only the
+        //    clock to the CPU that is gated. Clock is still available Systick
+        //    and NVIC peripherals is still available, which is what we want.
+        //
+        // 5. Do a DSB (If required, not needed in case of STM32F446RE)
+        //
+        // 6. Do a WFI (Go to sleep)
+        //
+        // 7. Exit critical section (handled by `atomic`)
```
Comment (ppannuto) [inline chips/stm32f446re/src/chip.rs]: This is a nice overview, but I'm wary of having so much kernel-specific stuff (e.g. interrupt bottom half queue) spelled out in a chip driver -- if it ever changes this doc is likely to bit rot :/

Consider moving some of this comment into the kernel loop itself, or the Tock documentation, and try to pare this down to just the STM32F-specific bits?

**Comment by rajivr** [comment]
> * Renaming kernel APIs needs to be its own PR.

I'll open a new PR for kernel API changes. 

I'll also keep the changes in this series, so as the discussion evolves on how
we want to do the top-half and bottom-half interrupt handler design and the
event dispatch mechanism, I can update this series.

There are two opposing requirements that can potentially make this challenging.

1. In order for me to reason about and explain the correctness of
   `INTERRUPT_EVENTS`, the design requires that I do not disable the interrupts
   at NVIC.
   
2. However, existing drivers for other boards, requires that we disable the
   interrupts at NVIC (using `generic_isr!`). There are no top-half handlers for
   existing drivers.

Any Kernel API change needs to provide a backward compatible way for board
maintainers to move existing drivers to a split driver model over time and I
don't have a good solution for that right now.

> * I am still opposed to adding a default `UserspaceKernelBoundary`

Okay. I'll remove it in the next update.

File: boards/kernel_layout.ld (hunk 4 above)
Comment (rajivr) [reply boards/kernel_layout.ld]: > Also, does this break the bundling mechanism right now? I think this needs to
> come after the `KEEP(*(.app*))` below

Okay I'll document the reason for `LONG(0)` and move it after `KEEP(*(.app*))`
in the next push.

Btw, where are we generating objects with `.app*` section which requires us to
have `KEEP(*(.app*))`?

**Comment by rajivr** [comment]
> That would be extremely unsafe, since there is no way to control the
> top-half's **concurrent** access to data structures.

@alevy Allow me to explain how concurrent access to data-structures is being
managed in this pull request. I believe @lthiery is also taking a similar
approach.

1. We need to identify the critical resources. That would be `INTERRUPT_EVENTS`
   for interrupt handling, and `EXTI_EVENTS` for external interrupt controller
   events. (Latter is STM32F4 specific)

2. Any write access to the critical resource must happen within an atomic
   closure. This is being done in the following ways -
   
   a. In `interrupt_context_isr!` (in `stm32f446re/src/lib.rs`), prior to
      modifying `INTERRUPT_EVENTS` and calling `$irq_fn` (which is the top-half
      interrupt handler), I do a `cpsid i` followed by `cpsie i`. This ensures
      that any state change we do in top-half interrupt handler happens
      atomically.
      
      The limitation here is that there are only a few things we can (and must
      do) in the top-half interrupt handler. Most of the time that would involve
      acknowledging to interrupt source, and capturing any state change that
      would be useful to us in the bottom-half.

      ```
      1:
        /* Due to the nested nature of NVIC interrupts, we need to disable   */
        /* interrupts before making changes to INTERRUPT_EVENTS. Once we are */
        /* done setting the appropriate bit, and calling the top half IRQ    */
        /* handler, we enable the interrupts again                           */
        cpsid i
      
        ldr r0, =INTERRUPT_EVENTS
        ldr r1, [r0, #0]
        orr r1, $0
        str r1, [r0, #0]
      
        mov r0, lr
        push {r0}"
        :
        : "{r2}"(0b1 << ($interrupt as u8))
        : : "volatile");
      
        // We do the mov, push, pop dance so lr is correctly saved and
        // restored.
        $irq_fn();
      
        asm!(
            "
        pop {r0}
        mov lr, r0
      
        cpsie i
      
        bx lr"
        :::: "volatile");
      ```

      If a global variable (more specifically a memory location) is modified in
      the interrupt context (top-half handler) or kernel context (bottom-half
      handler or regular kernel code), then its a critical resource.
 
   b. In order to make it easy to develop and reason about top-half interrupt
      handlers, I am using some conventions. I can explain those by the way of
      an example. If you look at `stm32f446re/src/dma1.rs`, you'll see that I
      introduce `interrupt_context` module to identify functions that are called
      from interrupt context.
      
      ```
      pub mod interrupt_context {
          // USART2_TX
          pub unsafe extern "C" fn irq_handler_dma1_stream6() {
                  super::Dma1Peripheral::USART2_TX
                  .stream()
                  .interrupt_context_finish_transfer();
          }

          [...]
      }
      ```

      Any code written inside `interrupt_context` needs to ensure that it is
      safe to be run inside an IRQ handler.
      
      The second convention is the use of `interrupt_context_XXX()` method. This
      is to indicate that this method is expected to be run inside an IRQ
      handler and needs to be IRQ safe.
      
   c. In kernel context access to critical resource is being managed within an
      atomic closure. This is done in `stm32f446re/src/interrupt_events.rs`.
      
      ```
      /// Checks if top half ISR have set any of the bits of `INTERRUPT_EVENTS`
      pub fn has_undispatched_interrupt_events() -> bool {
          let mut interrupt_events: u32 = 0;
      
          unsafe {
              atomic(|| {
                  interrupt_events = ptr::read_volatile(&INTERRUPT_EVENTS);
              });
          }
      
          interrupt_events != 0
      }
      
      /// Bottom half interrupt processing happens here. Using `atomic` closure, we
      /// load the value of `INTERRUPT_EVENTS` and `EXTI_EVENTS` into a stack
      /// variable. Also within the closure, we reset the value of `INTERRUPT_EVENTS`
      /// and `EXTI_EVENTS`. So, any top half ISRs that occur after exiting the
      /// closure will be processed in the next iteration of `kernel_loop`. Also
      /// before executing sleep using `wfi`, we check `INTERRUPT_EVENTS`. We sleep
      /// only if all the bottom halfs have been processed.
      pub fn dispatch_interrupt_events() {
          let mut interrupt_events: u32 = 0;
          let mut exti_events: u32 = 0;
      
          unsafe {
              atomic(|| {
                  interrupt_events = ptr::read_volatile(&INTERRUPT_EVENTS);
                  ptr::write_volatile(&INTERRUPT_EVENTS as *const u32 as *mut u32, 0);
      
                  exti_events = ptr::read_volatile(&exti::EXTI_EVENTS);
                  ptr::write_volatile(&exti::EXTI_EVENTS as *const u32 as *mut u32, 0);
              });
          }

          [...]

      }
      ```
      
While there is lot of review still required, I feel we should be able to
introduce split-driver model in Tock. 

2.b above is one way to manage top-half's concurrent access to data structures.

**Comment by lthiery** [comment]
> However, there is no consensus (and I'm personally still fairly skeptical) of using [interrupt events mechanism] more broadly. Moreover, it is definitely not intended to replace the vast majority of hardware interrupt handling.

I agree, when one can get away with it, keeping everything in the bottom-half is safer and simpler. However, there are countless situations where embedded systems _need_ to execute code in the interrupt routine. The quality and safety of drivers will be enhanced if we provide a canonical way of doing so rather than having every other project hack in some idiosyncratic approach.

Here are a few use-cases that I have coming up over the next few quarters where I do not think executing drivers in the bottom-half is sufficient:

- counting high frequency clicks / interrupt edges on a pin
- bit-banging obscure serial protocols (eg: SDI-12)
- high throughput ADC streams where I must mux from input to input after every sample

**Comment by rajivr** [comment]
> However, mixing in a new interrupt mechanism with a new board is a way
> too significant an architectural change to come as one unit. It looks
> like Tock's current interrupt handling mechanisms have some issues,
> and we should look to address those, but those changes have to be
> pulled into a separate PR.

I now understand the type-safety motivation behind the design of
`generic_isr()`. Also `generic_isr()` can be fixed by setting the
pending bit.

It is going to take me sometime to rework this pull request for
`generic_isr()`. I'll send an updated pull request after the holidays.

> Once/while we get the base STM merged, I hope you'll open a PR/Issue
> where we can start to hash out how to best properly support top-half
> interrupt handlers in Tock!)

Being able to do top-half interrupt handlers in a type-safe way, will
definitely be interesting. However, I don't know Rust well enough to
propose a way to do that! :-)

**Comment by alevy** [comment]
@bradjc is your "request for changes" still relevant?
